File: week1/4_Longest_Substring_Without_Repeating_Characters.py
# Longest Substring Without Repeating Characters
# Sliding Window
# Time Complexity: O(n)
# Space Complexity: O(n)

def lengthOfLongestSubstring(s: str) -> int:
  seen = set()
  l, r = 0, 0
  maxLength = 0

  for r in range(len(s)):
    while s[r] in seen:
      seen.remove(s[l])
      l += 1
    seen.add(s[r])
    maxLength = max(maxLength, r - l + 1)
  return maxLength

print(lengthOfLongestSubstring("abcabcbb"))
print(lengthOfLongestSubstring("bbbbb"))
print(lengthOfLongestSubstring("pwwkew"))

File: week1/4_Longest_Substring_Without_Repeating_Characters.ts
// Longest Substring Without Repeating Characters
// Sliding Window
// Time Complexity: O(n)
// Space Complexity: O(n)

const lengthOfLongestSubstring = (s: string): number => {
  const seen = new Set();
  let l = 0;
  let maxLength = 0;

  for (let r = 0; r < s.length; r++) {
    while (seen.has(s[r])) {
      seen.delete(s[l]);
      l++;
    }
    seen.add(s[r]);
    maxLength = Math.max(maxLength, r - l + 1);
  }
  return maxLength;
};
console.log(lengthOfLongestSubstring('abcabcbb'));
console.log(lengthOfLongestSubstring('bbbbb'));
console.log(lengthOfLongestSubstring('pwwkew'));
